sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java

```diff
@@ -436,23 +421,19 @@ public void processElement(ProcessContext context) throws Exception {
       @FinishBundle
       public void finishBundle(Context context) throws Exception {
         for (T record : batch) {
-          try {
-            PreparedStatement statement = connection.prepareStatement(query);
-            preparedStatementSetter.setParameters(record, statement);
-            try {
-              statement.executeUpdate();
-            } finally {
-              statement.close();
-            }
-          } catch (Exception e) {
-            LOGGER.error("Can't insert into database", e);
-          }
+          preparedStatementSetter.setParameters(record, preparedStatement);
+          preparedStatement.executeUpdate();
```

Comment:
Just noticed you're doing executeUpdate rather than executeBatch. This still executes the statements one-by-one, which is inefficient. You need to do addBatch (you can do it in processElement), and do executeBatch in finishBundle.